File: examples/route_and_reid/route_and_reid.cpp
// route_and_reid.cpp - native C++ port of route_and_reid.py: class-routed
// cascade + cross-camera re-ID, built directly on pycamtrt::Pipeline. THE
// showcase for Select (detection routing) and the Embedding family.
//
// Stage 1 ("detect") runs a COCO YOLOv8 detector; detections are ROUTED by
// class to two specialist siblings, each a Select -> Engine -> Postprocess
// 3-step layer: Select(classes={0}) (person) -> "person" (a resnet18
// penultimate-feature engine, Family::Embedding, 512-d vectors this binary
// cosine-matches ACROSS streams), and Select(classes={2,3,5,7},
// min_size=48) (vehicles) -> "vehicle" (a mobilenet ImageNet classifier,
// Family::Argmax). Honest caveats (same as the Python example): resnet18-
// on-ImageNet was never trained for re-ID (matches are illustrative),
// vehicle labels are loose ImageNet stand-ins, and the demo farm clip is
// an indoor scene (expect person matches, no vehicles).
//
// Usage:
//   route_and_reid URL [URL...]

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <deque>
#include <map>
#include <memory>
#include <string>
#include <vector>

#include "core/pipeline.h"

using pycamtrt::ChildOutput;
using pycamtrt::Family;
using pycamtrt::FrameResult;
using pycamtrt::LayerDesc;
using pycamtrt::Pipeline;
using pycamtrt::PipelineConfig;
using pycamtrt::StepDesc;
using pycamtrt::StepKind;
using pycamtrt::StreamDesc;
using pycamtrt::StreamInfo;

namespace {

const char* const kDetectEngine = "models/yolov8n_b1-16_fp16_sm86.engine";  // COCO, 80 cls
const char* const kEmbedEngine = "models/resnet18emb_b1-32_fp16_sm86.engine";  // [N,512]
const char* const kVehicleEngine = "models/mobilenet_v3s_b1-32_fp16_sm86.engine";

const float kNormOffset[3] = {-123.675f, -116.28f, -103.53f};
const float kNormScale[3] = {1.f / 58.395f, 1.f / 57.12f, 1.f / 57.375f};

const std::vector<int> kPersonClasses = {0};
const std::vector<int> kVehicleClasses = {2, 3, 5, 7};  // car, motorcycle, bus, truck
const float kVehicleMinSize = 48.f;

// Cosine similarity above which two person vectors count as "same person"
// across cameras - tune per model/content; 0.85 is a demo value.
const float kMatchThreshold = 0.85f;
const size_t kGalleryPerStream = 50;  // recent vectors kept per stream

const int kSkip = 2;
const int kMaxFrames = 600;

// A few well-known ImageNet-1k vehicle classes, for readable labels; any
// other label id prints as "imagenet_<id>" (see VehicleLabel below).
const std::map<int, std::string> kVehicleLabels = {
    {407, "ambulance"},   {468, "cab"},         {511, "convertible"},
    {555, "fire_engine"}, {569, "garbage_truck"}, {654, "minibus"},
    {656, "minivan"},     {675, "moving_van"},  {717, "pickup"},
    {734, "police_van"},  {751, "racer"},       {779, "school_bus"},
    {817, "sports_car"},  {864, "tow_truck"},   {867, "trailer_truck"},
    {874, "trolleybus"},
};

std::string VehicleLabel(int id) {
    auto it = kVehicleLabels.find(id);
    return it != kVehicleLabels.end() ? it->second : "imagenet_" + std::to_string(id);
}

// L2-normalizes `v` in place (same +1e-9 epsilon as the Python example).
void Normalize(std::vector<float>* v) {
    double sumsq = 0.0;
    for (float x : *v) sumsq += (double)x * x;
    const float norm = (float)std::sqrt(sumsq) + 1e-9f;
    for (float& x : *v) x /= norm;
}

float Dot(const std::vector<float>& a, const std::vector<float>& b) {
    float s = 0.f;
    for (size_t i = 0; i < a.size() && i < b.size(); i++) s += a[i] * b[i];
    return s;
}

PipelineConfig BuildConfig(const std::vector<std::string>& urls) {
    PipelineConfig cfg;
    cfg.streams.reserve(urls.size());
    for (const std::string& u : urls) cfg.streams.push_back(StreamDesc{u});
    cfg.skip = kSkip;
    cfg.max_frames = kMaxFrames;
    cfg.log = [](const std::string& s) { printf("%s\n", s.c_str()); };

    StepDesc engine_step;
    engine_step.kind = StepKind::Engine;
    engine_step.input = -1;
    engine_step.engine_path = kDetectEngine;
    const int engine_idx = (int)cfg.steps.size();
    cfg.steps.push_back(engine_step);

    StepDesc post_step;
    post_step.kind = StepKind::Postprocess;
    post_step.input = engine_idx;
    post_step.family = Family::YoloDetect;
    post_step.score_thresh = 0.4f;
    const int det_idx = (int)cfg.steps.size();
    cfg.steps.push_back(post_step);
    cfg.layers.push_back(LayerDesc{"detect", {engine_idx, det_idx}});

    // "person": Select(classes={0}) -> Engine(embedding) -> Postprocess.
    StepDesc person_sel;
    person_sel.kind = StepKind::Select;
    person_sel.input = det_idx;
    person_sel.sel_classes = kPersonClasses;
    const int person_sel_idx = (int)cfg.steps.size();
    cfg.steps.push_back(person_sel);

    StepDesc person_eng;
    person_eng.kind = StepKind::Engine;
    person_eng.input = person_sel_idx;
    person_eng.engine_path = kEmbedEngine;
    person_eng.color = 1;  // RGB
    for (int c = 0; c < 3; c++) {
        person_eng.norm_offset[c] = kNormOffset[c];
        person_eng.norm_scale[c] = kNormScale[c];
    }
    const int person_eng_idx = (int)cfg.steps.size();
    cfg.steps.push_back(person_eng);

    StepDesc person_post;
    person_post.kind = StepKind::Postprocess;
    person_post.input = person_eng_idx;
    person_post.family = Family::Embedding;
    const int person_post_idx = (int)cfg.steps.size();
    cfg.steps.push_back(person_post);
    cfg.layers.push_back(
        LayerDesc{"person", {person_sel_idx, person_eng_idx, person_post_idx}});

    // "vehicle": Select(classes={2,3,5,7}, min_size=48) -> Engine(argmax).
    StepDesc vehicle_sel;
    vehicle_sel.kind = StepKind::Select;
    vehicle_sel.input = det_idx;
    vehicle_sel.sel_classes = kVehicleClasses;
    vehicle_sel.sel_min_size = kVehicleMinSize;
    const int vehicle_sel_idx = (int)cfg.steps.size();
    cfg.steps.push_back(vehicle_sel);

    StepDesc vehicle_eng;
    vehicle_eng.kind = StepKind::Engine;
    vehicle_eng.input = vehicle_sel_idx;
    vehicle_eng.engine_path = kVehicleEngine;
    vehicle_eng.color = 1;
    for (int c = 0; c < 3; c++) {
        vehicle_eng.norm_offset[c] = kNormOffset[c];
        vehicle_eng.norm_scale[c] = kNormScale[c];
    }
    const int vehicle_eng_idx = (int)cfg.steps.size();
    cfg.steps.push_back(vehicle_eng);

    StepDesc vehicle_post;
    vehicle_post.kind = StepKind::Postprocess;
    vehicle_post.input = vehicle_eng_idx;
    vehicle_post.family = Family::Argmax;
    const int vehicle_post_idx = (int)cfg.steps.size();
    cfg.steps.push_back(vehicle_post);
    cfg.layers.push_back(
        LayerDesc{"vehicle", {vehicle_sel_idx, vehicle_eng_idx, vehicle_post_idx}});

    return cfg;
}

}  // namespace

int main(int argc, char** argv) {
    std::vector<std::string> urls(argv + 1, argv + argc);
    if (urls.empty()) urls.push_back("rtsp://localhost:8554/cam1");
    const int n_streams = (int)urls.size();

    std::unique_ptr<Pipeline> pipeline;
    try {
        pipeline.reset(new Pipeline(BuildConfig(urls)));
    } catch (const std::exception& e) {
        fprintf(stderr, "pipeline init failed: %s\n", e.what());
        return -1;
    }
    pipeline->Start();

    std::map<int, std::deque<std::vector<float>>> galleries;
    for (int i = 0; i < n_streams; i++) galleries[i];  // seed every stream, empty
    int frames = 0, persons = 0, vehicles = 0, matches = 0;
    std::map<std::string, int> vehicle_counts;
    float best_cos = 0.f;
    int best_s1 = -1, best_s2 = -1;

    for (;;) {
        FrameResult r;
        const Pipeline::PollStatus st = pipeline->Poll(&r, 500);
        if (st == Pipeline::PollStatus::Timeout) continue;
        if (st == Pipeline::PollStatus::Finished) break;

        frames++;
        // r.children[0] = "person" (Embedding: vectors), r.children[1] =
        // "vehicle" (Argmax: labels/label_scores) - same order as
        // cfg.layers[1:], mirroring Python's r.outputs["person"]/
        // r.outputs["vehicle"].
        const ChildOutput* person_out = r.children.size() > 0 ? &r.children[0] : nullptr;
        const ChildOutput* vehicle_out = r.children.size() > 1 ? &r.children[1] : nullptr;

        for (size_t i = 0; i < r.detections.size(); i++) {
            if (person_out && i < person_out->vectors.size() &&
                !person_out->vectors[i].empty()) {
                std::vector<float> v = person_out->vectors[i];
                Normalize(&v);
                persons++;
                for (auto& [sid, gal] : galleries) {
                    if (sid == r.stream_id || gal.empty()) continue;
                    float top = 0.f;
                    for (const auto& g : gal) top = std::max(top, Dot(g, v));
                    if (top >= kMatchThreshold) {
                        matches++;
                        if (top > best_cos) {
                            best_cos = top;
                            best_s1 = r.stream_id;
                            best_s2 = sid;
                        }
                        if (matches <= 10) {
                            printf("MATCH person s%d frame %d ~ s%d (cos %.3f)\n",
                                   r.stream_id, r.frame_no, sid, top);
                        }
                    }
                }
                auto& own = galleries[r.stream_id];
                own.push_back(v);
                if (own.size() > kGalleryPerStream) own.pop_front();
            }
            // "routed" == non-empty entry - a filtered-away detection
            // keeps this child's default (0, 0.0), same alignment
            // contract Select guarantees on the Python side.
            if (vehicle_out && i < vehicle_out->labels.size() &&
                vehicle_out->label_scores[i] != 0.0f) {
                vehicles++;
                vehicle_counts[VehicleLabel(vehicle_out->labels[i])]++;
            }
        }
    }

    std::vector<StreamInfo> per_stream;
    per_stream.reserve(n_streams);
    for (int i = 0; i < n_streams; i++) per_stream.push_back(pipeline->GetStreamInfo(i));
    pipeline.reset();

    printf("\n---- summary ----\n");
    printf("frames: %d  person crops embedded: %d  vehicle crops classified: %d\n",
           frames, persons, vehicles);
    printf("cross-camera person matches (cos >= %.2f): %d\n", kMatchThreshold, matches);
    if (best_s1 >= 0) printf("best match: s%d ~ s%d (cos %.3f)\n", best_s1, best_s2, best_cos);
    if (!vehicle_counts.empty()) {
        printf("vehicle types:\n");
        std::vector<std::pair<std::string, int>> sorted_counts(vehicle_counts.begin(),
                                                                 vehicle_counts.end());
        std::sort(sorted_counts.begin(), sorted_counts.end(),
                  [](const auto& a, const auto& b) { return a.second > b.second; });
        for (const auto& [name, count] : sorted_counts) printf("  %5d  %s\n", count, name.c_str());
    }
    for (int i = 0; i < n_streams; i++) {
        printf("s%d: decoded=%d reconnects=%d\n", i, per_stream[i].decoded,
               per_stream[i].reconnects);
    }
    return 0;
}

File: examples/route_and_reid/route_and_reid.py
#!/usr/bin/env python3
"""route_and_reid.py - class-routed cascade + cross-camera re-ID built on
pycamtrt. THE showcase for detection routing (``Select``) and the
``embedding`` family together.

Stage 1 ("detect") runs a COCO YOLOv8 detector over every stream. Its
detections are then ROUTED by class to two specialist siblings:

  - ``Select(classes={0})`` (person) -> stage "person": a resnet18
    penultimate-feature engine (``family="embedding"``) turns each person
    crop into a 512-d vector, and this script cosine-matches vectors
    ACROSS streams - "the person on cam1 is the person on cam2".
  - ``Select(classes={2,3,5,7}, min_size=48)`` (car/motorcycle/bus/truck,
    big enough to classify) -> stage "vehicle": a mobilenet ImageNet
    classifier (``family="argmax"``) labels the vehicle type.

Routing is exact and strictly reduces work: each sibling only ever
infers the crops its Select passes; every other detection keeps that
sibling's empty entry (``outputs[layer][i]`` stays aligned with
``detections[i]`` - one indexing rule, routed or not). See the Select
docstring in python/pycamtrt/__init__.py for the full contract.

Honest caveats: resnet18-on-ImageNet is a MECHANICS demo for re-ID - it
was never trained for person re-identification, so treat matches as
illustrative (a real deployment would export a re-ID checkpoint, e.g.
OSNet, through the same ``embedding`` family). Likewise the vehicle
labels are ImageNet classes (pickup, minivan, sports car, ...), a loose
stand-in for a purpose-trained vehicle classifier. And the demo farm clip
is an indoor webcam scene: expect person matches and no vehicles there -
point it at real street streams to see both branches light up.

Run it (from the repo root; full canonical docker incantation in
examples/read_plates/read_plates.py's docstring):

    PYTHONPATH=/workspace/build:/workspace/python \\
    python3 \\
    examples/route_and_reid/route_and_reid.py rtsp://localhost:8554/cam1 rtsp://localhost:8554/cam2

No cameras? tools/stream_farm/farm.sh serves a looping clip as N RTSP
streams (cd tools/stream_farm && ./farm.sh up 2).
"""
import sys
from collections import Counter, deque

import numpy as np

import pycamtrt

DETECT_ENGINE = "models/yolov8n_b1-16_fp16_sm86.engine"      # COCO, 80 cls
EMBED_ENGINE = "models/resnet18emb_b1-32_fp16_sm86.engine"   # [N,512] head
VEHICLE_ENGINE = "models/mobilenet_v3s_b1-32_fp16_sm86.engine"

IMAGENET_NORM = ((-123.675, -116.28, -103.53),
                 (1 / 58.395, 1 / 57.12, 1 / 57.375))

PERSON_CLASSES = {0}            # COCO ids - YOUR detector's ids, not ours
VEHICLE_CLASSES = {2, 3, 5, 7}  # car, motorcycle, bus, truck
VEHICLE_MIN_SIZE = 48           # skip crops too small to classify

# Cosine similarity above which two person vectors count as "same person"
# across cameras. Tune per model/content; 0.85 is a demo value.
MATCH_THRESHOLD = 0.85
GALLERY_PER_STREAM = 50         # recent vectors kept per stream

# A few well-known ImageNet-1k vehicle classes, for readable labels; any
# other label id prints as "imagenet_<id>".
VEHICLE_LABELS = {407: "ambulance", 468: "cab", 511: "convertible",
                  555: "fire_engine", 569: "garbage_truck", 654: "minibus",
                  656: "minivan", 675: "moving_van", 717: "pickup",
                  734: "police_van", 751: "racer", 779: "school_bus",
                  817: "sports_car", 864: "tow_truck", 867: "trailer_truck",
                  874: "trolleybus"}

SKIP = 2
MAX_FRAMES = 600


def build_pipeline(urls):
    streams = pycamtrt.Streams(urls)

    detect = pycamtrt.Layer("detect")
    eng = detect.add(pycamtrt.Engine(streams, DETECT_ENGINE))
    det = detect.add(pycamtrt.Postprocess(eng, family="yolo", score=0.4))

    person = pycamtrt.Layer("person")
    # The routing node: only class-0 (person) crops reach this engine.
    psel = person.add(pycamtrt.Select(det, classes=PERSON_CLASSES))
    peng = person.add(pycamtrt.Engine(psel, EMBED_ENGINE,
                                      norm=IMAGENET_NORM, color="rgb"))
    person.add(pycamtrt.Postprocess(peng, family="embedding"))

    vehicle = pycamtrt.Layer("vehicle")
    vsel = vehicle.add(pycamtrt.Select(det, classes=VEHICLE_CLASSES,
                                       min_size=VEHICLE_MIN_SIZE))
    veng = vehicle.add(pycamtrt.Engine(vsel, VEHICLE_ENGINE,
                                       norm=IMAGENET_NORM, color="rgb"))
    vehicle.add(pycamtrt.Postprocess(veng, family="argmax"))

    return pycamtrt.Pipeline(streams, layers=[detect, person, vehicle],
                             skip=SKIP, max_frames=MAX_FRAMES)


def main():
    urls = sys.argv[1:] or ["rtsp://localhost:8554/cam1"]

    galleries = {i: deque(maxlen=GALLERY_PER_STREAM)
                 for i in range(len(urls))}
    frames = persons = vehicles = matches = 0
    vehicle_counts = Counter()
    best_match = (0.0, None, None)

    with build_pipeline(urls) as pipe:
        for r in pipe:
            frames += 1
            for i, det in enumerate(r.detections):
                vec = r.outputs["person"][i] if i < len(
                    r.outputs["person"]) else []
                if vec:  # a routed person crop - 512 floats
                    persons += 1
                    v = np.asarray(vec, dtype=np.float32)
                    v /= np.linalg.norm(v) + 1e-9
                    # Cosine-match against OTHER streams' recent persons.
                    for sid, gal in galleries.items():
                        if sid == r.stream_id or not gal:
                            continue
                        sims = np.stack(gal) @ v
                        top = float(sims.max())
                        if top >= MATCH_THRESHOLD:
                            matches += 1
                            if top > best_match[0]:
                                best_match = (top, r.stream_id, sid)
                            if matches <= 10:
                                print(f"MATCH person s{r.stream_id} frame "
                                      f"{r.frame_no} ~ s{sid} "
                                      f"(cos {top:.3f})")
                    galleries[r.stream_id].append(v)
                pair = r.outputs["vehicle"][i] if i < len(
                    r.outputs["vehicle"]) else None
                if pair is not None and pair[1] != 0.0:  # routed vehicle
                    vehicles += 1
                    label = VEHICLE_LABELS.get(pair[0],
                                               f"imagenet_{pair[0]}")
                    vehicle_counts[label] += 1

        per_stream = [pipe.get_stream_info(i) for i in range(len(urls))]

    print("\n---- summary ----")
    print(f"frames: {frames}  person crops embedded: {persons}  "
          f"vehicle crops classified: {vehicles}")
    print(f"cross-camera person matches (cos >= {MATCH_THRESHOLD}): "
          f"{matches}")
    if best_match[1] is not None:
        print(f"best match: s{best_match[1]} ~ s{best_match[2]} "
              f"(cos {best_match[0]:.3f})")
    if vehicle_counts:
        print("vehicle types:")
        for label, count in vehicle_counts.most_common():
            print(f"  {count:5d}  {label}")
    for i, si in enumerate(per_stream):
        print(f"s{i}: decoded={si.decoded} reconnects={si.reconnects}")


if __name__ == "__main__":
    main()
